java/src/main/java/ai/onnxruntime/OrtEnvironment.java

```diff
@@ -0,0 +1,619 @@
+                    return new OnnxTensor(createString(OnnxRuntime.ortApiHandle, allocator.handle,(String)data), allocator.handle, info);
+                } else {
+                    return new OnnxTensor(createStringTensor(OnnxRuntime.ortApiHandle, allocator.handle, OrtUtil.flattenString(data), info.shape), allocator.handle, info);
+                }
+            } else {
+                if (info.shape.length == 0) {
+                    data = convertBoxedPrimitiveToArray(data);
+                }
+                return new OnnxTensor(createTensor(OnnxRuntime.ortApiHandle, allocator.handle, data, info.shape, info.onnxType.value), allocator.handle, info);
+            }
+        } else {
+            throw new IllegalStateException("Trying to create an OnnxTensor with a closed OrtAllocator.");
+        }
+    }
+
+    /**
+     * Create a tensor from a flattened string array.
+     * <p>
+     * Requires the array to be flattened in row-major order. Uses the default allocator.
+     * @param data The tensor data
+     * @param shape the shape of the tensor
+     * @return An OnnxTensor of the required shape.
+     * @throws OrtException Thrown if there is an onnx error or if the data and shape don't match.
+     */
+    public OnnxTensor createTensor(String[] data, long[] shape) throws OrtException {
+        return createTensor(defaultAllocator,data,shape);
+    }
+
+    /**
+     * Create a tensor from a flattened string array.
+     * <p>
+     * Requires the array to be flattened in row-major order.
+     * @param allocator The allocator to use.
+     * @param data The tensor data
+     * @param shape the shape of the tensor
+     * @return An OnnxTensor of the required shape.
+     * @throws OrtException Thrown if there is an onnx error or if the data and shape don't match.
+     */
+    OnnxTensor createTensor(OrtAllocator allocator, String[] data, long[] shape) throws OrtException {
+        if ((!closed) && (!allocator.isClosed())) {
+            TensorInfo info = new TensorInfo(shape, OnnxJavaType.STRING, OnnxTensorType.ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING);
+            return new OnnxTensor(createStringTensor(OnnxRuntime.ortApiHandle, allocator.handle, data, shape), allocator.handle, info);
+        } else {
+            throw new IllegalStateException("Trying to create an OnnxTensor on a closed OrtAllocator.");
+        }
+    }
+
+    /**
+     * Create an OnnxTensor backed by a direct FloatBuffer. The buffer should be in nativeOrder.
+     *
+     * If the supplied buffer is not a direct buffer, a direct copy is created tied to the lifetime
+     * of the tensor. Uses the default allocator.
+     * @param data The tensor data.
+     * @param shape The shape of tensor.
+     * @return An OnnxTensor of the required shape.
+     * @throws OrtException Thrown if there is an onnx error or if the data and shape don't match.
+     */
+    public OnnxTensor createTensor(FloatBuffer data, long[] shape) throws OrtException {
```

Comment:
Why does createTensor belong to the environment class? I would've expected it to be part of OnnxTensor class as a static factory.

Reply:
It needs to know about an allocator, and OrtAllocator is no longer a public class. The only instance of it lives inside OrtEnvironment at the moment, though it's written to allow use of custom allocators when it's decided that the Java API will support them. If the methods live on OnnxTensor then I need to check to see if the OrtEnvironment has been instantiated anyway, otherwise it could create a OrtTensor before the OrtEnv has been created.

Reply:
So conceptually the tensor needs an allocator and is not tied to the env. The fact that the allocator resides in the env is an implementation detail. We can still keep the allocator in the env in the following way: should we expose a method in the env to fetch the default allocator so that the user can pass it to the OnnxTensor::createTensor() method to create the appropriate tensor?

Reply:
We can do that, but at the moment OrtAllocator is closeable, but the close operation is a no-op on the default allocator. I'd prefer not to confuse the user by making them have to deal with an allocator that behaves a little different to all the rest. I could make the OnnxTensor::createTensor method accept either an environment or an allocator?

Reply:
It just feels unnatural to create a tensor off of the env class just because the allocator resides in the env. I would prefer the OnnxTensor::createTensor method accept an allocator instead.

Reply:
Ok I'll move them.